### **Who are the delayed flights?**
    the definition of flight delay is a flight that did not land on time.
   
    In cases where the flight was delayed at departure and arrived on time or arrived before the estimated landing time -
    this flight will not be considered a delayed flight.


### **Libraries**

In [2]:
import pandas as pd
import numpy as np
import datetime

In [3]:
dataset = pd.read_csv("data.csv")
dataset.head(5)

FL_DATE OP_CARRIER  OP_CARRIER_FL_NUM ORIGIN DEST  CRS_DEP_TIME  \
0  2018-01-01         UA               2429    EWR  DEN          1517   
1  2018-01-01         UA               2427    LAS  SFO          1115   
2  2018-01-01         UA               2426    SNA  DEN          1335   
3  2018-01-01         UA               2425    RSW  ORD          1546   
4  2018-01-01         UA               2424    ORD  ALB           630   

   DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  ...  CRS_ELAPSED_TIME  \
0    1512.0       -5.0      15.0      1527.0  ...             268.0   
1    1107.0       -8.0      11.0      1118.0  ...              99.0   
2    1330.0       -5.0      15.0      1345.0  ...             134.0   
3    1552.0        6.0      19.0      1611.0  ...             190.0   
4     650.0       20.0      13.0       703.0  ...             112.0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  \
0                250.0     225.0    1605.0            NaN            NaN   
1                 83.0      65.0     414.0            NaN            NaN   
2                126.0     106.0     846.0            NaN            NaN   
3                182.0     157.0    1120.0            NaN            NaN   
4                106.0      83.0     723.0            NaN            NaN   

  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  Unnamed: 27  
0       NaN             NaN                  NaN          NaN  
1       NaN             NaN                  NaN          NaN  
2       NaN             NaN                  NaN          NaN  
3       NaN             NaN                  NaN          NaN  
4       NaN             NaN                  NaN          NaN  

[5 rows x 28 columns]

### **List of features**
    FL_DATE = Date of the Flight
    OP_CARRIER = Airline Identifier
    OP_CARRIER_FL_NUM = Flight Number
    ORIGIN = Starting Airport Code
    DEST = Destination Airport Code
    CRS_DEP_TIME = Planned Departure Time
    DEP_TIME = Actual Departure Time
    DEP_DELAY = Total Delay on Departure in minutes
    TAXI_OUT = The time duration elapsed between departure from the origin airport gate and wheels off
    WHEELS_OFF = The time point that the aircraft's wheels leave the ground
    WHEELS_ON = The time point that the aircraft'ss wheels touch on the ground
    TAXI_IN = The time duration elapsed between wheels-on and gate arrival at the destination airport
    CRS_ARR_TIME = Planned arrival time
    ARR_TIME = Actual Arrival Time = ARRIVAL_TIME - SCHEDULED_ARRIVAL
    ARR_DELAY = Total Delay on Arrival in minutes
    CANCELLED = Flight Cancelled (1 = cancelled)
    CANCELLATION_CODE = Reason for Cancellation of flight: A - Airline/Carrier; B - Weather; C - National Air System; D - Security
    DIVERTED = Aircraft landed on different airport that the one scheduled
    CRS_ELAPSED_TIME = Planned time amount needed for the flight trip
    ACTUAL_ELAPSED_TIME = AIR_TIME+TAXI_IN+TAXI_OUT
    AIR_TIME = The time duration between wheels_off and wheels_on time
    DISTANCE = Distance between two airports
    CARRIER_DELAY = Delay caused by the airline in minutes
    WEATHER_DELAY = Delay caused by weather
    NAS_DELAY = Delay caused by air system
    SECURITY_DELAY = caused by security reasons
    LATE_AIRCRAFT_DELAY = Delay caused by security

### **Note :**
    as you can see , the size of the dataset is too big.
    In this notebook we will clean up the data and make sure to leave the relevant data for the model.

    First, we'll change the OP_CARRIER column which will contain the name of the airlines instead of their code.
    In addition with this column we use to exploring the data set.
    Second, we'll drop all the flights that been canceled 

In [45]:
dataset['OP_CARRIER'].replace({'UA':'United Airlines','AS':'Alaska Airlines','9E':'Endeavor Air','B6':'JetBlue Airways',     
                               'EV':'ExpressJet','F9':'Frontier Airlines','G4':'Allegiant Air','HA':'Hawaiian Airlines',
                               'MQ':'Envoy Air','NK':'Spirit Airlines','OH':'PSA Airlines','OO':'SkyWest Airlines',
                               'VX':'Virgin America','WN':'Southwest Airlines','YV':'Mesa Airline',
                               'YX':'Republic Airways','AA':'American Airlines','DL':'Delta Airlines'} , inplace = True)

dataset = dataset[(dataset['CANCELLED'] == 0)]
## drop OP_CARRIER , and save this column for Exploring the dataset
op_carrier = dataset.pop('OP_CARRIER')
dataset = dataset.drop(columns = ['Unnamed: 27','CANCELLED','CANCELLATION_CODE','DIVERTED','OP_CARRIER_FL_NUM','DEP_TIME', 'ARR_TIME'],axis = 1)

In [46]:
dataset.isna().sum()

FL_DATE                      0
ORIGIN                       0
DEST                         0
CRS_DEP_TIME                 0
DEP_DELAY                 4743
TAXI_OUT                     0
WHEELS_OFF                   0
WHEELS_ON                 2662
TAXI_IN                   2662
CRS_ARR_TIME                 0
ARR_DELAY                20456
CRS_ELAPSED_TIME             7
ACTUAL_ELAPSED_TIME      17858
AIR_TIME                 17858
DISTANCE                     0
CARRIER_DELAY          5744152
WEATHER_DELAY          5744152
NAS_DELAY              5744152
SECURITY_DELAY         5744152
LATE_AIRCRAFT_DELAY    5744152
dtype: int64

    As can be seen in the information above, the following columns:
    CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY : Contains almost 80% percent of empty values.
    (5744152 / 7213446) * 100 = 79.63

In [47]:
dataset = dataset.drop(columns = ['CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY'] , axis = 1)
dataset.head(5)

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  \
0  2018-01-01    EWR  DEN          1517       -5.0      15.0      1527.0   
1  2018-01-01    LAS  SFO          1115       -8.0      11.0      1118.0   
2  2018-01-01    SNA  DEN          1335       -5.0      15.0      1345.0   
3  2018-01-01    RSW  ORD          1546        6.0      19.0      1611.0   
4  2018-01-01    ORD  ALB           630       20.0      13.0       703.0   

   WHEELS_ON  TAXI_IN  CRS_ARR_TIME  ARR_DELAY  CRS_ELAPSED_TIME  \
0     1712.0     10.0          1745      -23.0             268.0   
1     1223.0      7.0          1254      -24.0              99.0   
2     1631.0      5.0          1649      -13.0             134.0   
3     1748.0      6.0          1756       -2.0             190.0   
4      926.0     10.0           922       14.0             112.0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  
0                250.0     225.0    1605.0  
1                 83.0      65.0     414.0  
2                126.0     106.0     846.0  
3                182.0     157.0    1120.0  
4                106.0      83.0     723.0

### **Dealing with Missing Values :**

In [48]:
dataset = dataset.dropna()
print(dataset.isna().sum())
dataset.shape

FL_DATE                0
ORIGIN                 0
DEST                   0
CRS_DEP_TIME           0
DEP_DELAY              0
TAXI_OUT               0
WHEELS_OFF             0
WHEELS_ON              0
TAXI_IN                0
CRS_ARR_TIME           0
ARR_DELAY              0
CRS_ELAPSED_TIME       0
ACTUAL_ELAPSED_TIME    0
AIR_TIME               0
DISTANCE               0
dtype: int64


(7071817, 15)

### **Change the date format :**

In [49]:
dataset['YEAR'] = pd.DatetimeIndex(dataset['FL_DATE']).year
dataset['DAY'] = pd.DatetimeIndex(dataset['FL_DATE']).day
dataset['MONTH'] = pd.DatetimeIndex(dataset['FL_DATE']).month
dataset = dataset.drop(['FL_DATE'],axis = 1)
dataset = dataset[['YEAR','MONTH','DAY','ORIGIN','DEST','CRS_DEP_TIME','DEP_DELAY','TAXI_OUT','WHEELS_OFF','WHEELS_ON','TAXI_IN',
                   'CRS_ARR_TIME','CRS_ELAPSED_TIME','ACTUAL_ELAPSED_TIME','AIR_TIME','DISTANCE','ARR_DELAY']]
dataset.head(5)

YEAR  MONTH  DAY ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  TAXI_OUT  \
0  2018      1    1    EWR  DEN          1517       -5.0      15.0   
1  2018      1    1    LAS  SFO          1115       -8.0      11.0   
2  2018      1    1    SNA  DEN          1335       -5.0      15.0   
3  2018      1    1    RSW  ORD          1546        6.0      19.0   
4  2018      1    1    ORD  ALB           630       20.0      13.0   

   WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  CRS_ELAPSED_TIME  \
0      1527.0     1712.0     10.0          1745             268.0   
1      1118.0     1223.0      7.0          1254              99.0   
2      1345.0     1631.0      5.0          1649             134.0   
3      1611.0     1748.0      6.0          1756             190.0   
4       703.0      926.0     10.0           922             112.0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  ARR_DELAY  
0                250.0     225.0    1605.0      -23.0  
1                 83.0      65.0     414.0      -24.0  
2                126.0     106.0     846.0      -13.0  
3                182.0     157.0    1120.0       -2.0  
4                106.0      83.0     723.0       14.0

### Binary Classification
    Our problem is classification, where a "0" will correspond to a flight being on time, and a "1" to a flight being delayed.

In [50]:
# change ARR_DELAY column: "0" will correspond to a flight being on time, and a "1" to a flight being delayed.
status = []
for value in dataset['ARR_DELAY']:
    if value <= 0:
        status.append(0)
    else:
        status.append(1)
dataset['ARR_DELAY'] = status

In [51]:
dataset.ARR_DELAY.value_counts(normalize = True)

0    0.644384
1    0.355616
Name: ARR_DELAY, dtype: float64

       As you can see  we have highly imbalanced data, as we there are only 35% rows with the value of 1.0 (Delay in flight).
       We will drop a significant amount of rows where our target variable is 0 (No delay in flight).
       And as a result, the size of our datta will be :(5073012,15)

In [52]:
# Split the data into positive and negative
positive_rows = dataset.ARR_DELAY == 1.0
data_pos = dataset.loc[positive_rows]
data_neg = dataset.loc[~positive_rows]

# Merge the balanced data
dataset = pd.concat([data_pos, data_neg.sample(n = len(data_pos))], axis = 0)
# Shuffle the order of data
dataset = dataset.sample(n = len(dataset)).reset_index(drop = True)
dataset.head(5)

YEAR  MONTH  DAY ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  TAXI_OUT  \
0  2018     10   11    GJT  SLC          1521       94.0      29.0   
1  2018      6   11    DCA  EWR           600       -7.0      19.0   
2  2018      8   31    SAT  FLL          1600       28.0       6.0   
3  2018      1   22    JFK  LAX          1910       -7.0      22.0   
4  2018      6   24    SEA  ORD          1735       99.0      12.0   

   WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  CRS_ELAPSED_TIME  \
0      1724.0     1810.0      3.0          1630              69.0   
1       612.0      649.0      9.0           712              72.0   
2      1634.0     2003.0      6.0          1950             170.0   
3      1925.0     2153.0     60.0          2244             394.0   
4      1926.0       56.0      4.0          2328             233.0   

   ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  ARR_DELAY  
0                 78.0      46.0     216.0          1  
1                 65.0      37.0     199.0          0  
2                161.0     149.0    1145.0          1  
3                410.0     328.0    2475.0          1  
4                226.0     210.0    1721.0          1

In [53]:
dataset.ARR_DELAY.value_counts(normalize = True)

0    0.5
1    0.5
Name: ARR_DELAY, dtype: float64

### One Hot encoding
       Because of the fact that Logistic Regression cannot use non-numeric data
       we use one hot encoding with pandas library to convert the non numerical columns to numerical values.
       It is important to mention that the size of the dataset will be significantly larger.
       Thus, the model will work with a data that contains 1% precent of the original dataset. (50297 examples)

In [54]:
dataset = pd.get_dummies(dataset , columns = ['ORIGIN','DEST'])
y_lable = dataset.pop('ARR_DELAY')
dataset = dataset.join(y_lable)
dataset.head(5)

YEAR  MONTH  DAY  CRS_DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  \
0  2018     10   11          1521       94.0      29.0      1724.0     1810.0   
1  2018      6   11           600       -7.0      19.0       612.0      649.0   
2  2018      8   31          1600       28.0       6.0      1634.0     2003.0   
3  2018      1   22          1910       -7.0      22.0      1925.0     2153.0   
4  2018      6   24          1735       99.0      12.0      1926.0       56.0   

   TAXI_IN  CRS_ARR_TIME  ...  DEST_VEL  DEST_VLD  DEST_VPS  DEST_WRG  \
0      3.0          1630  ...         0         0         0         0   
1      9.0           712  ...         0         0         0         0   
2      6.0          1950  ...         0         0         0         0   
3     60.0          2244  ...         0         0         0         0   
4      4.0          2328  ...         0         0         0         0   

   DEST_WYS  DEST_XNA  DEST_YAK  DEST_YNG  DEST_YUM  ARR_DELAY  
0         0         0         0         0         0          1  
1         0         0         0         0         0          0  
2         0         0         0         0         0          1  
3         0         0         0         0         0          1  
4         0         0         0         0         0          1  

[5 rows x 731 columns]

In [55]:
dataset.shape

(5029704, 731)

In [60]:
newdataset = dataset.sample(frac = 0.01)
newdataset.reset_index()
newdataset.head(5)

YEAR  MONTH  DAY  CRS_DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  \
2122964  2018      8    7          1340       -6.0       9.0      1343.0   
2867865  2018     10   29          1240       -8.0      13.0      1245.0   
2441876  2018      9   21          1934        0.0      25.0      1959.0   
4446265  2018     12   10          1945      102.0      11.0      2138.0   
111072   2018      3   11           820        5.0       9.0       834.0   

         WHEELS_ON  TAXI_IN  CRS_ARR_TIME  ...  DEST_VEL  DEST_VLD  DEST_VPS  \
2122964     1415.0      3.0          1440  ...         0         0         0   
2867865     1330.0      7.0          1350  ...         0         0         0   
2441876     2147.0     14.0          2159  ...         0         0         0   
4446265     2238.0      6.0          2105  ...         0         0         0   
111072       922.0      2.0           920  ...         0         0         0   

         DEST_WRG  DEST_WYS  DEST_XNA  DEST_YAK  DEST_YNG  DEST_YUM  ARR_DELAY  
2122964         0         0         0         0         0         0          0  
2867865         0         0         0         0         0         0          0  
2441876         0         0         0         0         0         0          1  
4446265         0         0         0         0         0         0          1  
111072          0         0         0         0         0         0          1  

[5 rows x 731 columns]

In [68]:
## check if the new data set is balanced
newdataset.ARR_DELAY.value_counts(normalize = True)
## save the cleanning data set (newdataset)
newdataset.to_csv("cleanningdata.csv")